cmd/transportRequestUploadCTS.go

```diff
@@ -0,0 +1,106 @@
+	WithConnection(transportrequest.CTSConnection)
+	WithApplication(transportrequest.CTSApplication)
+	WithNodeProperties(transportrequest.CTSNode)
+	WithTransportRequestID(string)
+	WithConfigFile(string)
+	WithDeployUser(string)
+}
+
+// GetAction ...
+func (provider *ActionProvider) GetAction() CTSUploadAction {
+	return provider.action
+}
+
+type transportRequestUploadCTSUtilsBundle struct {
+	*command.Command
+	*ActionProvider
+
+	// Embed more structs as necessary to implement methods or interfaces you add to transportRequestUploadUtils.
+	// Structs embedded in this way must each have a unique set of methods attached.
+	// If there is no struct which implements the method you need, attach the method to
+	// transportRequestUploadUtilsBundle and forward to the implementation of the dependency.
+}
+
+func newTransportRequestUploadCTSUtils() transportRequestUploadUtils {
+	utils := transportRequestUploadCTSUtilsBundle{
+		Command:        &command.Command{},
+		ActionProvider: &ActionProvider{action: &transportrequest.CTSUploadAction{}},
+	}
+	// Reroute command output to logging framework
+	utils.Stdout(log.Writer())
+	utils.Stderr(log.Writer())
+	return &utils
+}
+
+func transportRequestUploadCTS(config transportRequestUploadCTSOptions, telemetryData *telemetry.CustomData) {
+	// Utils can be used wherever the command.ExecRunner interface is expected.
+	// It can also be used for example as a mavenExecRunner.
+	utils := newTransportRequestUploadCTSUtils()
+
+	// For HTTP calls import  piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	// and use a  &piperhttp.Client{} in a custom system
+	// Example: step checkmarxExecuteScan.go
+
+	// Error situations should be bubbled up until they reach the line below which will then stop execution
+	// through the log.Entry().Fatal() call leading to an os.Exit(1) in the end.
+	err := runTransportRequestUploadCTS(&config, telemetryData, utils)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runTransportRequestUploadCTS(config *transportRequestUploadCTSOptions, telemetryData *telemetry.CustomData, utils transportRequestUploadUtils) error {
+
+	log.Entry().Debugf("Entering 'runTransportRequestUpload' with config: %v", config)
+
+	action := utils.GetAction()
+
+	action.WithConnection(transportrequest.CTSConnection{
+		Endpoint: config.Endpoint,
+		Client:   config.Client,
+		User:     config.Username,
+		Password: config.Password,
+	})
+	action.WithApplication(transportrequest.CTSApplication{
+		Name: config.ApplicationName,
+		Pack: config.AbapPackage,
+		Desc: config.Description,
+	})
+	action.WithNodeProperties(transportrequest.CTSNode{
+		DeployDependencies: config.DeployToolDependencies,
+		InstallOpts:        config.NpmInstallOpts,
+	})
+
+	action.WithTransportRequestID(config.TransportRequestID)
+	action.WithConfigFile(config.DeployConfigFile)
+	action.WithDeployUser(config.OsDeployUser)
+
+	return action.Perform(utils)
```

Comment:
it is a bit strange how the action comes out of `utils`, while the `Perform()` needs the utils again. I don't understand how the `Perform()` which needs a `ShellRunner` in the end gets this from the `utils` interface. But I guess it is my lack of golang knowledge that blocks me here.

Reply:
I separated the action from the utils. Please check if it looks more understandable now.